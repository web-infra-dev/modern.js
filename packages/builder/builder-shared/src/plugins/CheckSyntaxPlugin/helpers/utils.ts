import { CheckSyntaxExclude } from './type';

export function checkIsExcludeSource(
  path: string,
  exclude?: CheckSyntaxExclude,
) {
  if (!exclude) {
    return false;
  }

  const excludes = Array.isArray(exclude) ? exclude : [exclude];

  return excludes.some(reg => reg.test(path));
}
