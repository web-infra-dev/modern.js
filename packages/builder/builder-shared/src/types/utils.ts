export type JSONPrimitive = string | number | boolean | null | undefined;

export type JSONArray = Array<JSONValue>;

export type JSONObject = { [key: string]: JSONValue };

export type JSONValue = JSONPrimitive | JSONObject | JSONArray;

export type ArrayOrNot<T> = T | T[];

export type PromiseOrNot<T> = T | Promise<T>;

export type NodeEnv = 'development' | 'production' | 'test';

export type ChainedConfig<Config, Utils = unknown> = ArrayOrNot<
  | Config
  | (keyof Utils extends never
      ? (config: Config) => Config | void
      : (config: Config, utils: Utils) => Config | void)
>;

export type DeepReadonly<T> = keyof T extends never
  ? T
  : { readonly [k in keyof T]: DeepReadonly<T[k]> };

export type FileFilterUtil = (items: ArrayOrNot<string | RegExp>) => void;

export type SharedCompiledPkgNames =
  | 'pug'
  | 'sass'
  | 'less'
  | 'css-loader'
  | 'postcss-loader'
  | 'postcss-pxtorem'
  | 'postcss-flexbugs-fixes'
  | 'postcss-custom-properties'
  | 'postcss-initial'
  | 'postcss-page-break'
  | 'postcss-font-variant'
  | 'postcss-media-minmax'
  | 'postcss-nesting'
  | 'autoprefixer'
  | 'sass-loader'
  | 'less-loader'
  | 'node-loader'
  | 'babel-loader'
  | 'file-loader'
  | 'url-loader'
  | 'toml-loader'
  | 'yaml-loader'
  | 'assetsRetry.js'
  | 'resolve-url-loader';

export type CompilerTapFn<
  CallBack extends (...args: any[]) => void = () => void,
> = {
  tap: (name: string, cb: CallBack) => void;
};
