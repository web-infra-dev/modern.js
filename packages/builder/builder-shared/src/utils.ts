import type {
  SharedNormalizedConfig,
  BuilderTarget,
  SharedCompiledPkgNames,
} from './types';
import { join } from 'path';

export const createVirtualModule = (content: string) =>
  `data:text/javascript,${content}`;

export const removeLeadingSlash = (s: string): string => s.replace(/^\/+/, '');

export const addTrailingSlash = (s: string): string =>
  s.endsWith('/') ? s : `${s}/`;

export interface AwaitableGetter<T> extends PromiseLike<T[]> {
  promises: Promise<T>[];
}

/**
 * Make Awaitable.
 */
export const awaitableGetter = <T>(
  promises: Promise<T>[],
): AwaitableGetter<T> => {
  const then: PromiseLike<T[]>['then'] = (...args) =>
    Promise.all(promises).then(...args);
  return { then, promises };
};

export const isUseJsSourceMap = (config: SharedNormalizedConfig) => {
  const { disableSourceMap } = config.output || {};

  if (typeof disableSourceMap === 'boolean') {
    return !disableSourceMap;
  }

  return !disableSourceMap.js;
};

export const isUseCssSourceMap = (config: SharedNormalizedConfig) => {
  const { disableSourceMap } = config.output || {};

  if (typeof disableSourceMap === 'boolean') {
    return !disableSourceMap;
  }

  // If the disableSourceMap.css option is not specified, we will enable it in development mode.
  // We do not need CSS Source Map in production mode.
  if (disableSourceMap.css === undefined) {
    return process.env.NODE_ENV !== 'production';
  }

  return !disableSourceMap.css;
};

export const getSharedPkgCompiledPath = (
  packageName: SharedCompiledPkgNames,
) => {
  return join(__dirname, '../compiled', packageName);
};

// Determine if the string is a URL
export const isURL = (str: string) =>
  str.startsWith('http') || str.startsWith('//:');

export * as z from './zod';

export function isWebTarget(target: BuilderTarget | BuilderTarget[]) {
  return ['modern-web', 'web', 'web-worker'].some(t =>
    (Array.isArray(target) ? target : [target]).includes(t as BuilderTarget),
  );
}

export function resolvePackage(loader: string, dirname: string) {
  // Vitest do not support require.resolve to source file
  return process.env.VITEST
    ? loader
    : require.resolve(loader, { paths: [dirname] });
}
