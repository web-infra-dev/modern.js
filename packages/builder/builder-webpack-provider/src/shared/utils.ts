import { URLSearchParams } from 'url';
import {
  getBrowserslist,
  DEFAULT_BROWSERSLIST,
  DEFAULT_DATA_URL_SIZE,
} from '@modern-js/builder-shared';

import type Buffer from 'buffer';
import type { SomeJSONSchema } from '@modern-js/utils/ajv/json-schema';
import type { BuilderConfig, DataUriLimit } from '../types';

export async function getBrowserslistWithDefault(
  path: string,
  config: BuilderConfig,
) {
  if (config?.output?.overrideBrowserslist) {
    return config.output.overrideBrowserslist;
  }

  const result = await getBrowserslist(path);
  return result || DEFAULT_BROWSERSLIST;
}

/** Preserving the details of schema by generic types. */
export const defineSchema = <T extends SomeJSONSchema>(schema: T): T => schema;

export const getDataUrlLimit = (
  config: BuilderConfig,
  type: keyof DataUriLimit,
) => {
  const { dataUriLimit = {} } = config.output || {};

  if (typeof dataUriLimit === 'number') {
    return dataUriLimit;
  }

  switch (type) {
    case 'svg':
      return dataUriLimit.svg ?? DEFAULT_DATA_URL_SIZE;
    case 'font':
      return dataUriLimit.font ?? DEFAULT_DATA_URL_SIZE;
    case 'media':
      return dataUriLimit.media ?? DEFAULT_DATA_URL_SIZE;
    case 'image':
      return dataUriLimit.image ?? DEFAULT_DATA_URL_SIZE;
    default:
      throw new Error(`unknown key ${type} in "output.dataUriLimit"`);
  }
};

export function getDataUrlCondition(
  config: BuilderConfig,
  type: keyof DataUriLimit,
) {
  return (source: Buffer, { filename }: { filename: string }): boolean => {
    const queryString = filename.split('?')[1];

    if (queryString) {
      const params = new URLSearchParams(queryString);

      // If a request has query like "foo.png?__inline=false" or "foo.png?url",
      // no matter how small the file is, it will still be treated as a resource instead of being inlined
      const url = params.get('url');
      const legacyInline = params.get('__inline');
      if (legacyInline === 'false' || url !== null) {
        return false;
      }

      // If a request has query like "foo?inline" or "foo?__inline", no matter how big the file is,
      // it will be inlined instead of treated as a resource
      const inline = params.get('inline');
      if (inline !== null || legacyInline !== null) {
        return true;
      }
    }

    return source.length <= getDataUrlLimit(config, type);
  };
}

export async function stringifyConfig(config: unknown, verbose?: boolean) {
  const { default: WebpackChain } = await import(
    '../../compiled/webpack-5-chain'
  );

  // webpackChain.toString can be used as a common stringify method
  const stringify = WebpackChain.toString as (
    config: unknown,
    options: { verbose?: boolean },
  ) => string;

  return stringify(config as any, { verbose });
}
