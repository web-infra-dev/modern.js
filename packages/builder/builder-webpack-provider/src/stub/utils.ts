import fs, { PathLike } from 'fs';
import type { GlobbyOptions } from '@modern-js/utils';
import type webpack from 'webpack';

export interface GlobContentJSONOptions extends GlobbyOptions {
  maxSize?: number;
}

export const globContentJSON = async (
  paths: string | string[],
  options?: GlobContentJSONOptions,
) => {
  const { globby, fs } = await import('@modern-js/utils');
  const files = await globby(paths, options);
  let totalSize = 0;
  const maxSize = 1024 * (options?.maxSize ?? 4096);
  const ret: Record<string, string> = {};
  for await (const file of files) {
    const { size } = await fs.stat(file);
    totalSize += size;
    if (maxSize && totalSize > maxSize) {
      throw new Error('too large');
    }
    ret[file] = await fs.readFile(file, 'utf-8');
  }
  return ret;
};

export const filenameToGlobExpr = (file: PathLike) =>
  fs.statSync(file).isDirectory() ? `${file}/**/*` : file;

/**
 * Check if a file handled by specific loader.
 * @author yangxingyuan
 * @param {Configuration} config - The webpack config.
 * @param {string} loader - The name of loader.
 * @param {string}  testFile - The name of test file that will be handled by webpack.
 * @returns {boolean} The result of the match.
 */
export function matchLoader({
  config,
  loader,
  testFile,
}: {
  config: webpack.Configuration;
  loader: string;
  testFile: string;
}): boolean {
  if (!config.module?.rules) {
    return false;
  }
  return config.module.rules.some(rule => {
    if (
      typeof rule === 'object' &&
      rule.test &&
      rule.test instanceof RegExp &&
      rule.test.test(testFile)
    ) {
      return (
        Array.isArray(rule.use) &&
        rule.use.some(useOptions => {
          if (typeof useOptions === 'object' && useOptions !== null) {
            return useOptions.loader?.includes(loader);
          } else if (typeof useOptions === 'string') {
            return useOptions.includes(loader);
          }
          return false;
        })
      );
    }
    return false;
  });
}
