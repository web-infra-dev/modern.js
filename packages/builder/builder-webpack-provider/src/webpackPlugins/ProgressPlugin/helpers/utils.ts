export const clamp = (x: number, min: number, max: number) => {
  return Math.min(max, Math.max(min, x));
};
