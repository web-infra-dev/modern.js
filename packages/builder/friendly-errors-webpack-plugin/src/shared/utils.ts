import _ from '@modern-js/utils/lodash';
import { baseFormatter } from '../formatter';
import { ParsedError } from '../core/parse';
import { ErrorFormatter, ErrorTransformer } from './types';

export const cloneObject = <T>(error: T): T => {
  const cloned = Object.create(Object.getPrototypeOf(error));
  return _.merge(cloned, error);
};

export const cloneErrorObject = <T extends Error>(error: T): T => {
  const cloned = cloneObject(error);
  for (const k of Object.getOwnPropertyNames(error) as (keyof T)[]) {
    cloned[k] || (cloned[k] = error[k]);
  }
  return cloned;
};

export const transformError = (
  transformers: ErrorTransformer[],
  error: ParsedError,
): ParsedError => {
  return transformers.reduce(
    (error, transformer) => transformer(error) || error,
    error,
  );
};

export const formatError = (
  formatters: ErrorFormatter[] = [baseFormatter],
  error: ParsedError,
) => {
  let formatted = '';
  for (const formatter of formatters) {
    formatted = formatter(error) ?? formatted;
    if (formatted) {
      break;
    }
  }
  return formatted ?? baseFormatter(error);
};
