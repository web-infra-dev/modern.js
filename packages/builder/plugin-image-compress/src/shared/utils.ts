import assert from 'assert';
import _ from '@modern-js/utils/lodash';
import { FinalOptions, Options } from '../types';
import codecs from './codecs';

export const withDefaultOptions = (opt: Options): FinalOptions => {
  const options = _.isString(opt) ? { use: opt } : opt;
  const { defaultOptions } = codecs[options.use];
  const ret = { ...defaultOptions, ...options };
  assert('test' in ret);
  return ret;
};
