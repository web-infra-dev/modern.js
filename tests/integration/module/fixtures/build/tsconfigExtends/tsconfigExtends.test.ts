import path from 'path';
import { fs } from '@modern-js/utils';
import { runCli, initBeforeTest } from '../../utils';

initBeforeTest();

describe('tsconfig extends', () => {
  const fixtureDir = __dirname;
  it('should emit decorator metadata when extends from root tsconfig', async () => {
    const configFile = path.join(fixtureDir, './modern.config.ts');
    const { success, error } = await runCli({
      argv: ['build'],
      configFile,
      appDirectory: fixtureDir,
    });
    console.log(error);
    expect(success).toBeTruthy();
    const distFilePath = path.join(fixtureDir, 'dist/main.js');
    const content = await fs.readFile(distFilePath, 'utf-8');
    expect(content).toMatchSnapshot();
  });
});
