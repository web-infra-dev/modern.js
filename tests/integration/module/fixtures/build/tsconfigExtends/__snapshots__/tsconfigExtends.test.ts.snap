// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tsconfig extends should emit decorator metadata when extends from root tsconfig 1`] = `
""use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Person: () => Person
});
module.exports = __toCommonJS(src_exports);
function _ts_decorate(decorators, target, key, desc) {
  var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
    r = Reflect.decorate(decorators, target, key, desc);
  else
    for (var i = decorators.length - 1; i >= 0; i--)
      if (d = decorators[i])
        r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function _ts_metadata(k, v) {
  if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
    return Reflect.metadata(k, v);
}
function enhancer(name) {
  return function enhancer2(target) {
    target.prototype.name = name;
  };
}
var Person = class {
  constructor() {
    this.version = "1.0.0";
  }
};
Person = _ts_decorate([
  enhancer("module-tools"),
  _ts_metadata("design:type", Function),
  _ts_metadata("design:paramtypes", [])
], Person);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Person
});
"
`;
